docs/architecture/adr-021-abci-events.md

```diff
@@ -0,0 +1,47 @@
+# ADR 012: ABCI Events
+
+## Changelog
+
+- *2018-09-02* Remove ABCI errors component. Update description for events
+- *2018-07-12* Initial version
+
+## Context
+
+ABCI tags were first described in [ADR 002](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-002-event-subscription.md).
+They are key-value pairs that can be used to index transactions.
+
+Currently, ABCI messages return a list of tags to describe an
+"event" that took place during the Check/DeliverTx/Begin/EndBlock,
+where each tag refers to a different property of the event, like the sending and receiving account addresses.
+
+Since there is only one list of tags, recording data for multiple such events in
+a single Check/DeliverTx/Begin/EndBlock must be done using prefixes in the key
```

Comment:
Maybe I'm missing something fundamentally, but why is there a need for prefixing. Shouldn't the order for trying to search for things by tag be: Merkle tree of tags, `key: tag bytes, value: list of events`, then you get a merkle proof along with the entire tag set for that block? 

i.e. suppose my key is "slashing", then the value for that tag could be "[`Validator 0 slashed for downtime, other relevant details`, `Validator 2 slashed for not voting on proposal X`]"

Reply:
If you do it that way, you can't structure the value data. I might want to search for "all events where action = slashing and reason = downtime and validator = myValidator", for example, which isn't possible with this format (as I understand TM's current indexing model).

Reply:
Why can't you just query for all txs `validator = myValidator` and `slashing` filtering anything else client side?

Reply:
Because multiple validators can be slashed per `abci.RequestBeginBlock` - so you need to put them under different keys (or put them as a list under one key and somehow associate other parameters such as slash amount and reason).

Reply:
Is is possible to define mutliple tags with different values: `['validator=myValidator','validator=yourValidator']`?

Reply:
> If you do it that way, you can't structure the value data. I might want to search for "all events where action = slashing and reason = downtime and validator = myValidator", for example, which isn't possible with this format (as I understand TM's current indexing model).

With my proposal here: https://github.com/tendermint/tendermint/pull/2314#discussion_r214748281
that seems do-able. You can only proofs of exclude on one sub-field of a tag (e.g. only validator address, or only reason), but thats still better than no proofs of exclusion on an entire tag.

Reply:
Wait also, the flow as I understood it to previously occur doesn't requiring prefixing as well. You would simply require all events that are tagged some way meet a provided structure, and you could then do the same compound searching you described.

Reply:
See comment above. How do you suggest disambiguating which tags are associated with which others? Say two slashings happen in an `EndBlock` - if the resulting tags are `validator=validatorA, validator=validatorB, amount=2, amount=3`, we need a way to associate `validatorA` with `2` and `validatorB` with `3` - that's what I'm trying to get at.

Reply:
My understanding is the goal is to find all events that meet some criterion. The way I proposed doing this is to create a merkle tree where all the leafs are basically `TAG_NAME, TAG_TREE_HASH`. The tag `TAG_TREE_HASH` is then the root of a second merkle tree for all events meeting that tag. This enables us to do a proof of exclusion for an entire tag, and we can even compound this by sorting the values in a second manner to get the proof of exclusion on a single other tag (e.g. validator address, or amount). 

Suppose the search query is `slashing`. You would have to return the entire value tree of the `slashing` tag. This only has 1 merkle proof required (as your given all leaves of the slashing tag's tree)

Suppose the search query is `all matches for slashing AND amount=2 AND validator=validatorA`, and you want a proof that you got everything. Then to prove inclusion and no exclusion, I would return the smallest of the 3 tag trees. (i.e. either return the the entire `slashing` tree, `amount` tree, or `validator` tree. Any one of these would convince you that all events under that tag were included). Since the tags are sorted by name, if one of tags doesn't exist, its a rather simple proof of exclusion. This again has only one merkle proof required. (Though requires an entire sub-tree, however this is not at all a problem given how specific the trees are)

Suppose the search query is `all matches for slashing AND validator=validatorA`, and `validator` is the special second field which we sort by, that I mentioned above. Then we would return `merkle path to slashing` + `all slashing tree's merkle paths to correct answers` + `boundary elements to prove no exclusion`. 

I guess after describing this, I do think this scheme I'm proposing is better, but is sufficeintly complicated that it should be punted to postlaunch. (I also think we can use some more advanced cryptography to get even better proofs here, there are authenticated dictionaries) 

I still don't understand how we can do proofs of exclusion in the currently proposed scheme within the ADR, if someone could walk me through how you do a proof that no event was excluded for the search "All events under slashing" I would appreciate that.

Reply:
The current ADR doesnt say anything about how to do Merkle proofs - that's left for another issue: https://github.com/tendermint/tendermint/issues/1007#issuecomment-413917763. Would love it if you could help write up a proposal with the lite client requirements and some solutions.

My understanding is to some extent that the format in ABCI is independent of how we ultimately merklize everything.

Reply:
Oh woops, I didn't respond there mainly because I didn't understand the issue. This is starting to make sense to me now. I'd be glad to write up a proposal for that. 

I guess we could do a conversion layer between ABCI message type, and merkelization format, but that seems like complexity that might be annoying. Its also an O(total number of tags) conversion. (Not distinct tags), the conversion layer probably isn't a complexity concern given all of the other hashes that will be taking place.

Reply:
Note we already need to map from ABCI types to merkleable types because the ABCI responses may included non-deterministic data (ie. the log and info fields). Currently we're just merklizing the code+data, but the point of #1007 is to design a scheme for all the fields, ideally in a way that is more intelligent than just putting all the deterministic fields in a single struct (which wouldn't allow proof of exclusion, for instance). So it's worth having some doc to explain what's going on here and what the possible outcomes and tradeoffs are.

Reply:
Slashing doesn't happen during txs but does happen during BeginBlock, before the txs. BeginBlock returns tags so that clients can be notified of what "events" happened during the BeginBlock. If multiple validators get slashed during a single BeginBlock, we need a way to disentangle those multiple events (ie. each one has a validator address, an amount slashed, perhaps a reason for slashing, etc.)

Reply:
In theory that's possible, but would require significant change to the state machine, and is contrary to how we designed tendermint (ie. to auto detect and include evidence in blocks as a special type of data, distinct from txs).

Reply:
ok, then I retract my proposal